### PR TITLE
put overloads accepting WriteLogic

### PR DESCRIPTION
**Pull request by trejkaz**
Instead of using Consumer<T> which is Java 8 only and also forbids throwing an exception, I introduced an interface which can be used with lambdas in Java 8 anyway, but which will still work in earlier versions.

### CODE REVIEW

File: src/main/java/com/github/sardine/impl/SardineImpl.java [hunk 1]
```diff
@@ -765,13 +773,25 @@ public void put(String url, InputStream dataStream, String contentType, boolean
 	}
 
 	@Override
-	public void put(String url, InputStream dataStream, Map<String, String> headers) throws IOException {
+	public void put(String url, InputStream dataStream, Map<String, String> headers) throws IOException
+	{
+        this.put(url, dataStream, headerMapToList(headers));
+    }
+
+    @Override
```
Comment (lookfirst) [inline src/main/java/com/github/sardine/impl/SardineImpl.java]: tiny formatting issue.

**Comment by lookfirst** [comment]
Formatting should be tabs, not spaces. Other than that, it looks good.

**Comment by trejkaz** [comment]
Rats. I'll fix that. Pretty much every project I work on is the opposite. :)

**Comment by lookfirst** [comment]
I'm fine with both vars. just don't break backwards compat for that stuff cause it is hard to fix other dependent software.

**Comment by lookfirst** [comment]
Shoot.. wrong issue for that last comment that I deleted... please ignore. =)

**Comment by trejkaz** [comment]
I will note that I saw some other 4-space indents in SardineImpl. I fixed one of them which made the layout look particularly bad, but there are some other innocuous ones which might be worth doing in a separate commit.

**Comment by dkocher** [comment]
This should be handled outside of this library. See for example [DAVWriteFeature](https://github.com/iterate-ch/cyberduck/blob/master/webdav/src/main/java/ch/cyberduck/core/dav/DAVWriteFeature.java).

**Comment by trejkaz** [comment]
I took a look, but it isn't clear how I would solve this using their code as an example. They have the luxury of their data already being in a file, and I'm already using a temporary file to work around it today anyway, so it seems like there is nothing I can do to improve things.

**Comment by timmc** [comment]
I ran into this as well -- I have an encryption function that reads an input stream and writes to an output stream, and I wanted to open a WebDAV connection for that output stream. But I can't, because it only accepts an input stream. So I either have to do it in memory or write to a temp file, neither of which is great for large objects. Or, I have to spin up a thread to pump bytes from one place to another, also not great.

My other pluggable uploaders provide output streams -- file, S3, HTTP, etc. It seems to be the standard way of handling this sort of task in the wider JVM world. Would the maintainers be open to revisiting this?

**Pull request closed without merging**